conans/test/functional/generators/components/propagate_specific_components_test.py

```diff
@@ -0,0 +1,150 @@
+import textwrap
+import unittest
+
+from conans.test.utils.tools import TestClient
+
+
+class PropagateSpecificComponents(unittest.TestCase):
+    """
+        Feature: recipes can declare the components they are consuming from their requirements,
+        only those components should be propagated to their own consumers. If required components
+        doesn't exist, Conan will fail:
+         * Resolved versions/revisions of the requirement might provide different components or
+           no components at all.
+    """
+
+    top = textwrap.dedent("""
+        from conans import ConanFile
+
+        class Recipe(ConanFile):
+            name = "top"
+
+            def package_info(self):
+                self.cpp_info.components["cmp1"].libs = ["top_cmp1"]
+                self.cpp_info.components["cmp2"].libs = ["top_cmp2"]
+    """)
+
+    middle = textwrap.dedent("""
+        from conans import ConanFile
+
+        class Recipe(ConanFile):
+            name = "middle"
+            requires = "top/version"
+            def package_info(self):
+                self.cpp_info.requires = ["top::cmp1"]
+    """)
+
+    @classmethod
+    def setUpClass(cls):
```

Comment:
Is the test so slow that a setUpClass would make sense? Why not a setUp and make sure there are no cross-effects between tests?

Reply:
It was cleaner IMO, the `test_xxx` contains only the statements that are interesting to the test itself, while the setup is guaranteed to be the same for all of them. When a new generator implements this feature I hope we write a new `test_xxxx` function for it using the same recipes, this way it is harder to write something custom for that generator.

But I can move it to `setUp` or even copy it to every single test, we don't have anything written about this.

Reply:
I'm moving it to a `setUp` method